### PR TITLE
Add fuzzer targeting HCM and Router

### PR DESCRIPTION
**Pull request by robertfemmer**
This fuzzer targets the HTTP connection manager (HCM) with an installed Router and can model a HTTP connection with multiple Downstreams and Upstreams. The goal is to guard against life-time issues of streams held by the HCM.

### CODE REVIEW

**Comment by alyssawilk** [comment]
Adi's our fuzzing expert so throwing over to him.  He can land as it's test only.

File: test/common/http/hcm_router_fuzz_test.cc [hunk 1]
```diff
@@ -0,0 +1,575 @@
...
+    FuzzCluster* cluster = selectClusterByName(path);
+    if (!cluster) {
+      return nullptr;
+    }
+    return cluster->route_;
+  }
+
+  Upstream::ThreadLocalCluster* getThreadLocalCluster(absl::string_view name) {
+    FuzzCluster* cluster = selectClusterByName(name);
+    std::string sname(name.data(), name.size());
```
Comment (adisuissa) [inline test/common/http/hcm_router_fuzz_test.cc:423]: where is this used?

File: test/common/http/hcm_router_fuzz_test.cc [hunk 2]
```diff
@@ -0,0 +1,575 @@
...
+  StreamState response_state_{StreamState::PendingHeaders};
+  NiceMock<MockRequestEncoder> mock_request_encoder_;
+};
+
+// This class mocks an upstream cluster. It holds the set of
+// `FuzzUpstream` instances, which result from `FuzzDownstream`
+// requests
+class FuzzCluster {
+public:
+  FuzzCluster(FuzzConfig& cfg, const char* path, bool internal_redirect_policy_enabled,
```
Comment (adisuissa) [inline test/common/http/hcm_router_fuzz_test.cc:241]: Please avoid using `char*`

File: test/common/http/hcm_router_fuzz_test.cc [hunk 3]
```diff
@@ -0,0 +1,575 @@
...
+              bool cross_scheme_redirect_allowed, bool allows_early_data_for_request)
+      : cfg_(cfg), name_(path), path_(path), mock_route_(new Router::MockRoute()),
```
Comment (adisuissa) [inline test/common/http/hcm_router_fuzz_test.cc:243]: This is a test code, so I guess it is less important, but `mock_route_` is allocated and never freed. Can this use a unique_ptr instead?

File: test/common/http/hcm_router_fuzz_test.cc [hunk 4]
```diff
@@ -0,0 +1,575 @@
...
+      Buffer::OwnedImpl buf(data);
+      s->sendData(buf, end_stream);
+    }
+  }
+
+  void sendTrailers(uint32_t stream, const test::fuzz::Headers& response_trailers) {
+    FuzzUpstream* s = select(stream);
+    if (s) {
+      auto trailers = std::make_unique<Http::TestResponseTrailerMapImpl>(
+          Fuzz::fromHeaders<Http::TestResponseTrailerMapImpl>(response_trailers, {}, {"status"}));
```
Comment (adisuissa) [inline test/common/http/hcm_router_fuzz_test.cc:302]: Is `status` needed as part of the trailers?

**Review by adisuissa — COMMENTED** [review]
Thanks!
Left a few comments, but overall LGTM.
It would be interesting to see which code areas this fuzzer will cover.

File: test/common/http/hcm_router_fuzz_test.cc (hunk 4 above)
Comment (robertfemmer) [reply test/common/http/hcm_router_fuzz_test.cc]: no, removed.

File: test/common/http/hcm_router_fuzz_test.cc (hunk 3 above)
Comment (robertfemmer) [reply test/common/http/hcm_router_fuzz_test.cc]: `mock_route_` will be used to initialize the `shared_ptr route_`, which will eventually free `mock_route_`. I agree this is awkward, but the alternative is to cast `route_` back to a `MockRoute`. I have now opted to initialize these two in the constructor body in order to make it more readable.

File: test/common/http/hcm_router_fuzz_test.cc (hunk 2 above)
Comment (robertfemmer) [reply test/common/http/hcm_router_fuzz_test.cc]: changed to  `absl::string_view`

File: test/common/http/hcm_router_fuzz_test.cc (hunk 1 above)
Comment (robertfemmer) [reply test/common/http/hcm_router_fuzz_test.cc]: removed sname

File: test/common/http/hcm_router_fuzz_test.cc [hunk 5]
```diff
@@ -0,0 +1,575 @@
...
+      internal_redirect_policy_enabled_ = internal_redirect;
+      cross_scheme_redirect_allowed_ = allow_cross_scheme;
+      allows_early_data_for_request_ = allow_early_data;
+    }
+    maintenance_ = maintenance;
+  }
+
+  FuzzConfig& cfg_;
+  std::string name_;
+  const absl::string_view path_;
```
[the remark this answers is not in the document]
Comment (robertfemmer) [reply test/common/http/hcm_router_fuzz_test.cc]: fixed

File: test/common/http/hcm_router_fuzz_test.cc [hunk 6]
```diff
@@ -0,0 +1,575 @@
...
+namespace Envoy {
+namespace Http {
+// The set of singleton objects that hold state across multiple fuzz
+// cases
+class Harness;
+class FuzzConfig;
+class FuzzClusterManager;
+static std::unique_ptr<Harness> harness = nullptr;
+static std::unique_ptr<FuzzConfig> hcm_config = nullptr;
+static std::unique_ptr<FuzzClusterManager> cluster_manager = nullptr;
```
[the remark this answers is not in the document]
Comment (robertfemmer) [reply test/common/http/hcm_router_fuzz_test.cc]: I moved all of these except the harness into other objects.

**Comment by phlax** [comment]
@adisuissa i think this is wating on next review

/wait-any

**Review by adisuissa — COMMENTED** [review]
Thanks!
Overall looks good. The main things to consider is the use of non-smart pointers and the static harness.

@krajshiva any comments?

**Review by adisuissa — COMMENTED** [review]
Thanks!
Overall looks good. The main things to consider is the use of non-smart pointers and the static harness.

@krajshiva any comments?

**Comment by KBaichoo** [comment]
/wait

I think this is waiting on changes from Adi's review @robertfemmer .

File: test/common/http/hcm_router_fuzz_test.cc [hunk 7]
```diff
@@ -0,0 +1,567 @@
...
+  NiceMock<LocalInfo::MockLocalInfo> local_info_;
+  NiceMock<Server::MockOverloadManager> overload_manager_;
+  NiceMock<Network::MockReadFilterCallbacks> filter_callbacks_;
+  std::unique_ptr<ConnectionManagerImpl> hcm_;
+
+  bool closed_{false};
+  std::vector<FuzzDownstreamPtr> streams_;
+};
+
+static std::unique_ptr<Harness> harness = nullptr;
```
[the remark this answers is not in the document]
Comment (robertfemmer) [reply test/common/http/hcm_router_fuzz_test.cc:588]: Allocating, initializing and freeing the harness on every run impacts performance severely. I would argue the benefit is substantial.

[the remark this answers is not in the document]
Comment (adisuissa) [reply test/common/http/hcm_router_fuzz_test.cc:588]: We have had a few cases where a false-positive was generated in fuzz tests due to dirty state, and I'd like to have a way to test without the persistent mode if possible.
Can you add a flag (macro) that disables the use of static structures if needed?

[the remark this answers is not in the document]
Comment (robertfemmer) [reply test/common/http/hcm_router_fuzz_test.cc:600]: Ok, done.

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!

**Pull request merged**